src/main/scala/translating/ReadELFLoader.scala

```diff
@@ -6,16 +6,56 @@ import util.ILLoadingConfig
 
 import scala.jdk.CollectionConverters.*
 
+/**
+ * https://refspecs.linuxfoundation.org/elf/elf.pdf
+ */
+
+enum ELFSymType:
+  case NOTYPE /* absolute symbol or similar */
+  case SECTION /* memory section */
+  case FILE
+  case OBJECT
+  case FUNC /* code function */
+
+
+enum ELFBind:
+  case LOCAL  /* local to the translation unit */
+  case GLOBAL /* global to the program */
+  case WEAK  /* multiple versions of symbol may be exposed to the linker, and the last definition is used. */
+
+enum ELFVis:
+  case HIDDEN
+  case DEFAULT
+
+enum ELFNDX:
+  case Section(num: Int) /* Section containing the symbol */
+  case UND /* Undefined */
+  case ABS /* Absolute, unaffected by relocation */
+
+case class ELFSymbol(num: Int,  /* symbol number */
+  value: BigInt, /* symbol address */
+  size: Int,  /* symbol size (bytes) */
+  etype: ELFSymType,  
+  bind: ELFBind,  
+  vis: ELFVis,
+  ndx: ELFNDX, /* The section containing the symbol */
+  name: String)
+
 object ReadELFLoader {
-  def visitSyms(ctx: SymsContext, config: ILLoadingConfig): (Set[ExternalFunction], Set[SpecGlobal], Map[BigInt, BigInt], Int) = {
+  def visitSyms(ctx: SymsContext, config: ILLoadingConfig): (List[ELFSymbol], Set[ExternalFunction], Set[SpecGlobal], Map[BigInt, BigInt], Int) = {
     val externalFunctions = ctx.relocationTable.asScala.flatMap(r => visitRelocationTableExtFunc(r)).toSet
     val relocationOffsets = ctx.relocationTable.asScala.flatMap(r => visitRelocationTableOffsets(r)).toMap
-    val globalVariables = ctx.symbolTable.asScala.flatMap(s => visitSymbolTable(s)).toSet
     val mainAddress = ctx.symbolTable.asScala.flatMap(s => getFunctionAddress(s, config.mainProcedureName))
+
+    val symbolTable = ctx.symbolTable.asScala.flatMap(s => visitSymbolTable(s)).toList
+    val globalVariables = (symbolTable.collect {
+      case ELFSymbol(num, value, size, ELFSymType.OBJECT, ELFBind.GLOBAL, ELFVis.DEFAULT, _, name) =>  SpecGlobal(name, size * 8, None, value)
```

Comment:
Removing the `allowedChars` check means that variables with illegal characters in Boogie are now added to the .bpl output in some cases. In practice this `allowedChars` check just meant that references to external objects, such as the `stack_chk_guard`, were not included in the set of `SpecGlobal`s. A more robust solution would probably be to exclude symbol table entries with ndx `UND` rather than relying on them containing an illegal character such as `@`.

Reply:
Ah yeah removing that was a mistake, ndx UND sounds like a good fix.